src/Assets/ugui-mvvm/Editor/TypeSuggestionProvider.cs

```diff
@@ -0,0 +1,105 @@
+﻿using AutoSuggest;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace uguimvvm
+{
+    public class TypeSuggestion : Suggestion
+    {
+        public Type Type { get; }
+
+        public TypeSuggestion(string errorMessage)
+            : base(errorMessage)
+        { }
+
+        public TypeSuggestion(Type type, string searchString)
+            : base(type.FullName, type.FullName, type.FullName.IndexOf(searchString, StringComparison.OrdinalIgnoreCase), searchString.Length)
+        {
+            Type = type;
+        }
+    }
+
+    public class TypeSuggestionProvider : AsyncSuggestionProvider
+    {
+        private IEnumerable<Type> _types;
+
+        public Type SelectedType { get; private set; } = null;
+        public bool SelectedTypeIsValid { get; protected set; } = false;
+
+        public override async Task<IEnumerable<Suggestion>> GetSuggestionsAsync(string currentValue, bool isFocused, CancellationToken cancellationToken)
+        {
+            IList<TypeSuggestion> results = null;
+
+            SelectedTypeIsValid = false;
+            SelectedType = null;
+
+            await Task.Run(() =>
+            {
+                if (_types == null)
+                {
+                    var typeQuery = AppDomain.CurrentDomain.GetAssemblies().SelectMany(a =>
+                    {
+                        try
+                        {
+                            return a.GetTypes();
+                        }
+                        catch (Exception)
+                        {
+                            return new Type[] { };
+                        }
+                    });
+                    // Calling ToList forces the query to execute this one time, instead of executing every single time "types" is enumerated.
+                    _types = typeQuery.ToList();
+                }
+
+                results = _types.AsParallel()
+                    .WithCancellation(cancellationToken)
+                    .Select((t) => new TypeSuggestion(t, currentValue))
+                    .Where((s) => s.DisplayTextMatchIndex >= 0)
+                    .OrderBy(opt => opt.DisplayTextMatchIndex)
+                    .ThenBy(opt => opt.Value.Length)
+                    .ThenBy(opt => opt.Value)
+                    .ToList();
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (results.Count == 1 && results[0].Value == currentValue)
+                {
+                    SelectedType = results[0].Type;
+                }
+                else
+                {
+                    SelectedType = null;
+                }
+                SelectedTypeIsValid = true;
```

Comment:
should this be false in the latter part of the condition above ?

Reply:
No.  The "IsValid" really is meant to say whether it is loading or has loaded.  So after setting the value to either null or not null, it is then valid.  I'll add a comment for clarity and I'm open to suggestions for a better name for the variable.